### ISSUE TITLE
bug: objects don't downsync after delete and recreate of placement object

### ISSUE THREAD

**Issue by effi-ofer** [kubestellar/kubestellar#1710]
### Describe the bug

Creating a placement + deployment on WDS downsync the deployment to the WEC.  If we then delete placement followed by create placement, then the deployment won't always downsync.  It does not help if we add a sleep(10) or a get deployment from WEC and confirm there is none between the delete and create placement.

### Steps To Reproduce
See a comment below on how to reproduce from scratch.  But the gist of it is captured in the following instructions:
```
$ kubectl --context wds1 delete placements nginx-placement; kubectl --context cluster1 get deployment -n nginx; kubectl --context wds1 create -f - <<EOF
apiVersion: control.kubestellar.io/v1alpha1
kind: Placement
metadata:
  name: nginx-placement
spec:
  clusterSelectors:
  - matchLabels: {"location-group":"edge"}
  downsync:
  - objectSelectors:
    - matchLabels: {"app.kubernetes.io/name":"nginx"}
EOF
$ kubectl --context cluster1 get deployments -n nginx


```
### Expected Behavior

Deployment gets downsynced.

**Comment by effi-ofer** [comment]
This is holding back PR #1713

**Comment by pdettori** [comment]
is there any useful info in the kubestellar controller manager log ? also, can you check what happens with manifestworks on imbs1 for cluster1 namespace?

**Comment by effi-ofer** [comment]
To reproduce from the command line do the following:
```
cd test/e2e/common
./cleanup.sh
./setup-kubestellar.sh
kubectl --context wds1 apply -f - <<EOF
apiVersion: v1
kind: Namespace
metadata:
  labels:
    app.kubernetes.io/name: nginx
  name: nginx
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: nginx
  labels:
    app.kubernetes.io/name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: public.ecr.aws/nginx/nginx:latest
        ports:
        - containerPort: 80
EOF
```
Then run in a loop:
```
kubectl --context wds1 delete placements nginx-placement; kubectl --context cluster1 get deployment -n nginx; kubectl --context wds1 create -f - <<EOF
apiVersion: control.kubestellar.io/v1alpha1
kind: Placement
metadata:
  name: nginx-placement
spec:
  clusterSelectors:
  - matchLabels: {"location-group":"edge"}
  downsync:
  - objectSelectors:
    - matchLabels: {"app.kubernetes.io/name":"nginx"}
EOF
kubectl --context cluster1 get deployments -n nginx
```

**Comment by effi-ofer** [comment]
@pdettori looking at the ks controller I see it logging a delivery of the deployment to the WEC.  ```
2024-02-06T09:49:09Z    INFO    Trying to delete manifest       {"manifest name": "v1-namespace--nginx", "namespace": "cluster1", "for placement": "nginx-placement"}
2024-02-06T09:49:09Z    INFO    Trying to delete manifest       {"manifest name": "v1-namespace--nginx", "namespace": "cluster2", "for placement": "nginx-placement"}
2024-02-06T09:49:09Z    INFO    Trying to delete manifest       {"manifest name": "appsv1-deployment-nginx-nginx-deployment", "namespace": "cluster1", "for placement": "nginx-placement"}
2024-02-06T09:49:09Z    INFO    Trying to delete manifest       {"manifest name": "appsv1-deployment-nginx-nginx-deployment", "namespace": "cluster2", "for placement": "nginx-placement"}
E0206 09:49:09.904537       1 controller.go:328] object cache.ObjectName{Namespace:"cluster1", Name:"appsv1-deployment-nginx-nginx-deployment"} in work queue no longer exists
E0206 09:49:09.905900       1 controller.go:328] object cache.ObjectName{Namespace:"cluster2", Name:"appsv1-deployment-nginx-nginx-deployment"} in work queue no longer exists
Matched key control.kubestellar.io/v1alpha1/Placement
2024-02-06T09:49:10Z    INFO    Matched {"object": "[] namespace/nginx", "for placement": "nginx-placement"}
2024-02-06T09:49:10Z    INFO    Matched {"object": "[nginx] deployment.apps/nginx-deployment", "for placement": "nginx-placement"}
2024-02-06T09:49:10Z    INFO    Delivering      {"object": "[] namespace/nginx", "to clusters": ["cluster1", "cluster2"]}
2024-02-06T09:49:10Z    INFO    Delivering      {"object": "[nginx] deployment.apps/nginx-deployment", "to clusters": ["cluster1", "cluster2"]}```

Checking the manifestworks I see the following:
```
$ kubectl --context imbs1 get manifestworks -A
NAMESPACE   NAME                                       AGE
cluster1    addon-addon-status-deploy-0                20m
cluster2    addon-addon-status-deploy-0                20m
cluster1    appsv1-deployment-nginx-nginx-deployment   9m44s
cluster2    appsv1-deployment-nginx-nginx-deployment   9m44s
```

**Comment by nirrozenbaum** [comment]
@effi-ofer might be useful to see also the agent log and look for an error message about namespace creation.

**Comment by effi-ofer** [comment]
Good suggestion @nirrozenbaum.  The OCM work agent logs the following 
```
E0206 10:02:46.458261       1 manifestwork_controller.go:226] Reconcile work appsv1-deployment-nginx-nginx-deployment fails with err: deployments.apps "nginx-deployment" is forbidden: unable to create new content in namespace nginx because it is being terminated
E0206 10:02:46.458328       1 base_controller.go:270] "ManifestWorkAgent" controller failed to sync "appsv1-deployment-nginx-nginx-deployment", err: deployments.apps "nginx-deployment" is forbidden: unable to create new content in namespace nginx because it is being terminated
```
Perhaps the delete placement finalizer did not wait for the namespace deletion on the WEC.  If that's true, I should be able to always always succeed (i.e. stop reproducing the failure) by confirming that the namespace has been deleted on the WEC before issuing the create placement.  Checking.

**Comment by nirrozenbaum** [comment]
ok. so now we know for sure that the problem is that namespace creation by the agent fails.
now the question is why it happens. this could happen due to the following two reasons:
1. there might be a finalizer on the namespace itself which prevents it from being deleted; OR
2. there might be a finalizer on one of the objects inside the namespace, which prevents the namespace deletion.

another step towards the solution would be to understand which finalizer prevents the deletion. then we can try to understand who should remove the finalizer and why this bug happens.

**Comment by effi-ofer** [comment]
Not exactly.  The namespace creation fails because that same namespace already existed and was in the **terminating** state.  The question is why didn't the delete placement wait for the namespace to be deleted?  Or maybe the create placement should recognize that the namespace creation failed and retry a few times before finally failing.

**Comment by nirrozenbaum** [comment]
k8s controller controllers design patten is based on eventual consistency.
placement controller don't even have any relationship with the WEC cluster (and the specific namespace). 
we can't expect the controller to wait for things to happen. we should achieve eventual consistency by the controller pattern.

in case the e2e test uses number of retries before failing, I think it should not specify number of retries but to use a ```kubectl wait for a condition``` command.   
 
btw, I also thinks you **should not** change the e2e to wait for completion of ns deletion on the WEC.
what we see here is a real bug (not a very critical bug, but a real one).
the system should be able to handle such cases where namespace is deleted and immediately recreated.
the e2e test should not avoid this bug, but instead to validate it is not happening anymore.

**Comment by pdettori** [comment]
I wonder if this is an issue that can be replicated just on OCM w/o KubeStellar whatsoever - e.g. manually create the manifest with the namespace and the manifest with NGNIX, apply to the 2 manifests  to a cluster namespace on imbs1 and delete and recreate both and see what happens. We may also try to do this with one single manifest that contains both namespace and deployment, which is close to what the transport code does, and see if that makes things better.

**Comment by effi-ofer** [comment]
That's a worth while test Paolo. While I set this up I have another question for you @nirrozenbaum - would the new transport/placement code change this behaviour?  Is it worth while trying the steps outlined in this issue on your new PR?

**Comment by nirrozenbaum** [comment]
if that is an issue with OCM (which we will know after your test), then I assume it would still happen in the newer version. 
even if it doesn’t happen with vanilla OCM, I strongly recommend using kubectl wait in tests. as explained, k8s is based on eventual consistency and using wait gives us what we need to test the system.

**Comment by effi-ofer** [comment]
I run the following which deletes and re-creates the namespace manifestwork. I can see that the manifestwork object has a finalizer that waits on the resource deletion on the managed cluster via the appliedmanifestwork object.  Consequently, if these delete + recreate are run serially then the create always blocks on the delete and the namespace is eventually created. This is consistent with the manifest documentation https://open-cluster-management.io/scenarios/deploy-kubernetes-resources/. If I run these in parallel, then either the create happens after the delete or the create fails because the manifestwork object still exists.  
```
kubectl --context imbs1 delete  manifestwork -n cluster1 v1-namespace--nginx 
kubectl --context imbs1 create -f m.yaml 
kubectl --context cluster1 get ns | grep nginx
```
Looking at the kubestellar code, I can see that the placement has a finalizer and checking the code it spins over the manifestwork objects and delete them before getting cleared.  If the namespace already exists on the managed cluster, the code still creates the manifestwork object.

**Comment by pdettori** [comment]
@nirrozenbaum is this an issue that the new transport code is addressing ?

**Comment by nirrozenbaum** [comment]
I think this is addressed in the new code.
we make sure we first deleted all wrapped objects and only then remove the finalizer.
also there should be no problem if the create fails cause this will requeue the `Binding` which will result in a retry (that eventually will succeed).
I will sync today with @effi-ofer and test the exact same scenario using the new transport controller.
will update with my results in this issue.

**Comment by effi-ofer** [comment]
Thanks Nir.  I'll gladly work with you.  I thought the placement delete issues a delete on the manifest objects which would in turn wait on their own finalizer which will delete the object on the managed cluster and then remove the wrapped object.  But I haven't stepped through this code to confirm it works as expected.  Did you find an issue there that you changed in your PR?

**Comment by effi-ofer** [comment]
Nir and I tried the following successfully: 
Running with the new transport decoupled layer without kubestellar, Nir created a deployment and a PlacementDecision object on the WSD.  We observed the deployment transported to the WEC.  We then in a loop deleted the PlacementDecision and created it again, each time verifying that the deployment got deleted and recreated at the WEC.  

At this point, the suspect is the placement controller, but since Maroon is changing it in PR #1681 and it's close to merging, I will wait for 1681 to be merged before investigating further.

**Comment by pdettori** [comment]
@effi-ofer thank you for the update, sounds like the new code may solve this issue.

**Comment by effi-ofer** [comment]
I confirmed this scenario no longer reproduce when running with the PR #1780 code base.  We should reconfirm and close this issue once the PR merges.

**Comment by nirrozenbaum** [comment]
@effi-ofer @ezrasilvera is this bug fixed in main after merging pluggable transport?
